File: 4485/4485.cpp
/*
Title   : BOJ_4485 [녹색 옷 입은 애가 젤다지?]
Author  : Hoseok Lee
Date    : 2023/03/21
https://www.acmicpc.net/problem/4485
https://github.com/hoshogi
*/

#include <iostream>
#include <queue>
#define INF 1e9
using namespace std;

int n;
int dr[4] = {-1, 0, 1, 0};
int dc[4] = {0, 1, 0, -1};
int board[125][125];
int result[125][125];

void bfs() {
    queue<pair<int, int>> q;
    result[0][0] = board[0][0];
    q.push({0, 0});
    
    while (!q.empty()) {
        int r = q.front().first;
        int c = q.front().second;
        int count = result[r][c];
        q.pop();
        
        for (int i = 0; i < 4; i++) {
            int tempR = r + dr[i];
            int tempC = c + dc[i];
            
            if (tempR < 0 || tempR >= n || tempC < 0 || tempC >= n)
                continue;
                
            int tempCount = count + board[tempR][tempC];
            if (tempCount < result[tempR][tempC]) {
                result[tempR][tempC] = tempCount;
                q.push({tempR, tempC});
            }
        }
    }
}

int main(void) {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    int count = 1;

    while (true) {
        cin >> n;
        
        if (n == 0)
            break;
        
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                cin >> board[i][j];
        for (int i = 0; i < n; i++)
            fill_n(result[i], n, INF);
        bfs();
        cout << "Problem " << count++ << ": " << result[n - 1][n - 1] << '\n';
    }
    return 0;
}

File: 4485/4485.java
/*
Title   : BOJ_4485 [녹색 옷 입은 애가 젤다지?]
Author  : Hoseok Lee
Date    : 2024/03/27

https://www.acmicpc.net/problem/4485
https://github.com/hoshogi
*/

import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    private static int len, result;
    private static int[][] graph, costs;
    private static int[] dr = {-1, 0, 1, 0};
    private static int[] dc = {0, 1, 0, -1};
    private static final int MAX = 1_000_000_000;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder answer = new StringBuilder();

        int count = 0;

        while (true) {
            count++;
            len = Integer.parseInt(br.readLine());
            if (len == 0) {
              break;
            }

            graph = new int[len][len];
            for (int i = 0; i < len; i++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                for (int j = 0; j < len; j++) {
                    graph[i][j] = Integer.parseInt(st.nextToken());
                }
            }

            costs = new int[len][len];
            for (int i = 0; i < len; i++) {
                Arrays.fill(costs[i], 0, len, MAX);
            }
            costs[0][0] = graph[0][0];
            result = MAX;
            dfs(0, 0, costs[0][0]);
            answer.append("Problem ").append(count).append(": ").append(result).append("\n");
        }
        bw.write(answer.toString());
        bw.flush();
        bw.close();
    }

    private static void dfs(int r, int c, int cost) {
        if (r == len - 1 && c == len - 1) {
            result = Math.min(result, cost);
            return;
        }

        for (int i = 0; i < 4; i++) {
            int tempR = r + dr[i];
            int tempC = c + dc[i];

            if (tempR < 0 || tempR >= len || tempC < 0 || tempC >= len)
                continue;
            if (costs[tempR][tempC] > cost + graph[tempR][tempC]) {
                costs[tempR][tempC] = cost + graph[tempR][tempC];
                if (costs[tempR][tempC] <= result)
                    dfs(tempR, tempC, costs[tempR][tempC]);
            }
        }
    }
}
